include/nuttx/clk/clk-provider.h

```diff
@@ -0,0 +1,277 @@
+
+#ifndef __INCLUDE_NUTTX_CLK_CLK_PROVIDER_H
+#define __INCLUDE_NUTTX_CLK_CLK_PROVIDER_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/list.h>
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifdef CONFIG_CLK
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define CLK_SET_RATE_GATE               0x01
+#define CLK_SET_PARENT_GATE             0x02
+#define CLK_SET_RATE_PARENT             0x04
+#define CLK_SET_RATE_NO_REPARENT        0x08
+#define CLK_GET_RATE_NOCACHE            0x10
+#define CLK_NAME_IS_STATIC              0x20
+#define CLK_PARENT_NAME_IS_STATIC       0x40
+#define CLK_IS_CRITICAL                 0x80
+
+#define CLK_GATE_SET_TO_DISABLE         0x01
+#define CLK_GATE_HIWORD_MASK            0x02
+
+#define CLK_DIVIDER_ONE_BASED           0x01
+#define CLK_DIVIDER_HIWORD_MASK         0x02
+#define CLK_DIVIDER_ROUND_CLOSEST       0x04
+#define CLK_DIVIDER_READ_ONLY           0x08
+#define CLK_DIVIDER_MAX_HALF            0x10
+#define CLK_DIVIDER_DIV_NEED_EVEN       0x20
+#define CLK_DIVIDER_POWER_OF_TWO        0x40
+#define CLK_DIVIDER_MINDIV_OFF          8
+#define CLK_DIVIDER_MINDIV_MSK          0xff00
+
+#define CLK_FRAC_MUL_NEED_EVEN          0x01
+#define CLK_FRAC_DIV_DOUBLE             0x02
+
+#define CLK_MULT_ONE_BASED              0x01
+#define CLK_MULT_ALLOW_ZERO             0x02
+#define CLK_MULT_HIWORD_MASK            0x04
+#define CLK_MULT_MAX_HALF               0x08
+#define CLK_MULT_ROUND_CLOSEST          0x10
+
+#define CLK_MUX_HIWORD_MASK             0x01
+#define CLK_MUX_READ_ONLY               0x02
+#define CLK_MUX_ROUND_CLOSEST           0x04
+
+#define CLK_PHASE_HIWORD_MASK           0x01
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+#undef EXTERN
+#if defined(__cplusplus)
+#  define EXTERN extern "C"
+extern "C"
+{
+#else
+#  define EXTERN extern
+#endif
+
+struct clk_rate
+{
+  const char *name;
+  uint32_t   rate;
+};
+
+struct clk
+{
+  const char              *name;
+  const struct clk_ops    *ops;
+  struct clk              *parent;
+  uint8_t                 num_parents;
+  uint8_t                 new_parent_index;
+  uint8_t                 enable_count;
+  uint8_t                 flags;
+  uint32_t                rate;
+  uint32_t                new_rate;
+  struct clk              *new_parent;
+  struct clk              *new_child;
+  void                    *private_data;
+  struct list_node        children;
+  struct list_node        node;
+  const char              *parent_names[0];
+};
+
+struct clk_ops
+{
+  int       (*enable)(struct clk *clk);
```

Comment:
```suggestion
  CODE int       (*enable)(struct clk *clk);
```
Missing `CODE` qualifier on pointer to code, required for generic code on NuttX